Produce deterministic pyc files in base_library.zip

achow101 opened the pull request:
For the pyc files in base_library.zip, produce the deterministic pyc files as described in PEP-552.

The reason for this is that I found when using pyenv and trying to do deterministic builds with multiple machines, builds were not deterministic due to the pyc file in base_library.zip having different timestamps that corresponded to the time that pyenv installed python.

I was unable to figure out how to get this to work when the hashes are checked so this uses the unchecked mode.

htgoebel commented:
Thanks for this pull-request. Reproducibility is important for me, so I'm happy you spotted this issue.

Regarding the code: We have similar code somewhwre else in PyInstaller (to make the archives reproducible9, I assume this is in`PyInstaller/archive/writers.py`. To avoid duplicate code please refactor both occurrences into a new function. `PyInstaller.building.utils` might be a good place. Thanks.

When doing this please keep in mind that we still need to support Python 2.7, thus the existing code would be the way to go.

Also please submit a [**changelog entry**](https://pyinstaller.readthedocs.io/en/latest/development/changelog-entries.html) so our users can learn about your change. This one should go into the `feature` or `core` category.

achow101 commented:
I looked through `PyInstaller/archive/writers.py` and I found that it is calling `fake_pyc_timestamp` in `PyInstaller/building/utils.py`. This seems to be the similar code that you were talking about, but I'm not sure that this is refactor-able. It looks like the archive builder is using pyc files that already exist and modifying them, while the base_library builder is creating pyc files from scratch.

achow101 commented:
I've updated this to be on top of the latest `develop` branch, added the changelog entry, updated the commit to (hopefully) fit the commit message style, and (hopefully) made it work with Python 2.7.

htgoebel commented on PyInstaller/depend/utils.py:
```diff
@@ -80,9 +80,15 @@ def create_py3_base_library(libzip_filename, graph):
+                                fc.write(struct.pack('<I', 0b1))
+                                with open(mod.filename, 'rb') as fs:
+                                    source_bytes = fs.read()
+                                import importlib.util
```
Please move this to the top (wrapped by try/except). We try to avoid imports in the code.

achow101 replied:
Done

htgoebel commented on news/4096.core.rst:
```diff
@@ -0,0 +1 @@
+Now uses PEP 552 for the `.pyc` files in `base_library.zip` when using Python 3.7.
```
This is not quite correct, since PEP 552 is followed since #3626 already. This is more like "uses hash based …"

You can also write `:pep:`552`` which will create a reference to the resp. PEP.

achow101 replied:
Done

htgoebel commented on PyInstaller/depend/utils.py:
```diff
@@ -80,9 +80,15 @@ def create_py3_base_library(libzip_filename, graph):
-                                fc.write(struct.pack('<I', 0))
-                            fc.write(struct.pack('<II', timestamp, size))
+                                # one means hash based but don't check the hash
+                                fc.write(struct.pack('<I', 0b1))
```
Since you are using binary notation, what about specifying the "check_source" flag, too (`0b01`) and adjust the comment to something like "0b01 means hash based but don#t check the hash"?

achow101 replied:
Done

htgoebel commented:
Thanks agin for quickly updating the pull-request (and the pull-request itself, of course).
